fix_includes.py

```diff
@@ -1339,87 +1339,158 @@ def _GetFirstNamespaceLevelReorderSpan(file_lines):
+      # header guard, which is (by definition) the only kind of #ifdef
+      # that we can be inside and still considered at the "top level".
+      if line_info.type in (_COMMENT_LINE_RE,
+                            _BLANK_LINE_RE,
+                            _INCLUDE_RE,
+                            _HEADER_GUARD_RE,
+                            _HEADER_GUARD_DEFINE_RE,
+                            _PRAGMA_ONCE_LINE_RE):
+        continue
 
-    elif line_info.type in (_NAMESPACE_START_ALLMAN_RE,
-                            _NAMESPACE_START_MIXED_RE):
-      # Trim any possible comments on namespace to properly build
-      # the namespace level spans
-      m = simple_allman_namespace_re.match(line_info.line)
-      if not m:
-        break
-      namespace_prefix += ('namespace %s' % m.group(1).strip())
+      # If we're a 'contentful' line such as a (non-header-guard) #ifdef, add
+      # to the ifdef depth.  If we encounter #endif, reduce the ifdef depth.
+      # Only keep track of namespaces when ifdef depth is 0
+      elif line_info.type == _IF_RE:
+        ifdef_depth += 1
 
-    elif line_info.type == _NAMESPACE_CONTINUE_ALLMAN_MIXED_RE:
-      # Append to the simplified Allman namespace
-      namespace_prefix += ' { '
+      elif line_info.type == _ELSE_RE:
+        continue
 
-    elif line_info.type == _FORWARD_DECLARE_RE:
-      # If we're not in a namespace, keep going.  Otherwise, this is
-      # just the situation we're looking for!
-      if namespace_prefix:
-        return (namespace_prefix, line_info.reorder_span)
+      elif line_info.type == _ENDIF_RE:
+        ifdef_depth -= 1
+
+      elif ifdef_depth != 0:
+        continue # skip lines until we're outside of an ifdef block
+
+      # Build the simplified namespace lists and namespace depths.  When
+      # any new namespace is encountered, add the namespace to the list
+      # using the next line to cover namespaces without forward declares.
+      # When a forward declare is found, add the concatenated namespaces
+      # to the list.  Once a contentful line (None) has been found, add
+      # the final namespace if we're still inside one and stop searching.
+      elif line_info.type == _NAMESPACE_START_RE:
+        for namespace in _GetNamespaceNames(line_info.line):
+          if not namespace:
+            namespace_prefixes.append('namespace { ')
+          else:
+            namespace_prefixes.append('namespace %s { ' % namespace)
+          namespace_depth += 1
+        
+        namespace_reorder_spans[''.join(namespace_prefixes)] = (
+          line_number+1, line_number+1)
 
-    else:
-      # We should have handled all the cases above!
-      assert False, ('unknown line-info type',
-                     _LINE_TYPES.index(line_info.type))
+      elif line_info.type == _NAMESPACE_START_ALLMAN_RE:
+        for namespace in _GetNamespaceNames(line_info.line):
+          if not namespace:
```

Comment:
This conditional code keeps coming back. I wonder if you should define a local function to do this, something like:

    def _NamespaceDecl(nsname):
      namespace = ['namespace']
      if nsname:
        namespace += [nsname]
      return ' '.join(namespace)

That way the inline code would be simpler:

    for namespace in _GetNamespaceNames(line_info.line):
      namespace_prefix += _NamespaceDecl(namespace)

I'm not really happy with the name `_NamespaceDecl`, but I think this would flatten a lot of the already pretty complex code here.

Reply:
I'm not too sure about this one.  There's a lot of repeating if/else statements that are a bit hard to digest but every case needs to deal with anonymous namespaces as well as regular namespaces in all three types of situations.  In some cases, we can just add right to the namespace_prefixes while in others, we need to build the pending namespace prefix.  Some cases need to add the { while others need to wait for the actual bracket to add it.  I'd have to add an additional argument to have the bracket appended to the return type I believe.  I can still add it but I just don't know if it'll be cleaner in the end, given all the special cases.